### PR TITLE
Implement an IVersionManager to avoid passing function pointers in Core.ts

### PR DESCRIPTION
**Pull request by mudiali**
- The VersionManager.ts now implements IVersionManager. 
- All components which require the 'versioned' objects now take in the IVersionManager instead of taking the function pointers as before.
- Also added tests for VersionManager.ts

### CODE REVIEW

File: tests/core/BatchScheduler.spec.ts [hunk 1]
```diff
@@ -2,12 +2,18 @@ import * as retry from 'async-retry';
 import BatchScheduler from '../../lib/core/BatchScheduler';
 import MockBatchWriter from '../mocks/MockBatchWriter';
 import MockBlockchain from '../mocks/MockBlockchain';
+import MockVersionedProvider from '../mocks/MockVersionManager';
```
Comment (thehenrytsai) [inline tests/core/BatchScheduler.spec.ts]: Nit:
```suggestion
import MockVersionManager from '../mocks/MockVersionManager';
```

File: tests/core/versions/testversion1/BatchWriter.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import IBatchWriter from '../../../../lib/core/interfaces/IBatchWriter';
```
Comment (thehenrytsai) [inline tests/core/versions/testversion1/BatchWriter.ts]: Nit: Consider following more of javascript convension of using hyphens between words.
`testversion1` -> `test-version-1` or maybe avoiding it altogether to `0.0.1`

Comment (mudiali) [reply tests/core/versions/testversion1/BatchWriter.ts]: Renaming.

File: tests/core/BatchScheduler.spec.ts (hunk 1 above)
Comment (mudiali) [reply tests/core/BatchScheduler.spec.ts]: Good catch … I missed this during the name changes.

**Pull request merged**